### PR TITLE
Added support for diagonal runs

### PR DESCRIPTION
**Pull request by Blaze-Leo**
Added support for diagonal run.

Hey Mack, please see if the changes are to your satisfaction. And don't worry about the readme, at the end i will update that.

### CODE REVIEW

File: src/Direction.cpp [hunk 1]
```diff
@@ -8,6 +8,10 @@ const QVector<Direction>& DIRECTIONS() {
         Direction::EAST,
         Direction::SOUTH,
         Direction::WEST,
+        Direction::NORTHEAST,
+        Direction::NORTHWEST,
+        Direction::SOUTHEAST,
+        Direction::SOUTHWEST,
```
Comment (mackorone) [inline src/Direction.cpp:14]: See my comments on #21 regarding a new enum.

File: src/Mouse.cpp [hunk 2]
```diff
@@ -36,12 +37,12 @@ Mouse::Mouse() {
     Coordinate centerOfMass = Coordinate::Cartesian(
         Distance::Meters(0.06),
         Distance::Meters(0.06)
-    );
+        );
```
Comment (mackorone) [inline src/Mouse.cpp:40]: Nitpick: please remove these whitespace changes

File: src/Mouse.cpp [hunk 3]
```diff
@@ -70,14 +71,31 @@ void Mouse::reset() {
...
 }
 
 void Mouse::teleport(const Coordinate& translation, const Angle& rotation) {
     m_currentTranslation = translation;
     m_currentRotation = rotation;
 }
 
+QPair<int, int> Mouse::getCurrentHalfDiscretizedTranslation() const {
```
Comment (mackorone) [inline src/Mouse.cpp:82]: I still have a bunch of questions about this method - see comments on #21

**Review by mackorone — CHANGES_REQUESTED** [review]
@Blaze-Leo I stopped reviewing after looking at the first files, once it became apparent that you haven't yet incorporated all of my feedback from #21. Please do that before resubmitting.

**Comment by Blaze-Leo** [comment]
Using a new `enum` for the `SemiDirections` is not possible as the variable `m_startingDirection` is set with data type `Direction`. Creating another variable just for the `SemiDirections` is not advisable.

**Comment by Blaze-Leo** [comment]
I have completed all the other changes you wanted.

**Comment by mackorone** [comment]
@Blaze-Leo thanks, and apologies for the radio silence! A quick update from my end: I've been playing around with your proposed changes for the past week or so. I may decide to incorporate them myself if I deem that easier/faster/safer (still TBD). My ETA is sometime within the next week or so. Stay tuned!

**Comment by Blaze-Leo** [comment]
Feel free to do what suits you, glad to help you in whatever decision you make.

**Comment by mackorone** [comment]
@Blaze-Leo, I finally finished incorporating some of your proposed changes into the simulator. In particular, it now supports edge-based movement, including diagonals. Please play around with the new API and let me know your thoughts. So far I've only updated the [mackorone/mms-python](https://github.com/mackorone/mms-python) template to include the new API methods - I'll update the other templates once I get some feedback from you.

**Comment by Blaze-Leo** [comment]
@mackorone  Which one is the diagonal movement command?

**Comment by Blaze-Leo** [comment]
Hey! I just started checking and i think your and my perspective of the position of the ir sensor in the mouse are different.
Suppose I am at this position
pos
According to your code the wall is sensed so the ir sensors must be here
irmack
But my code that I used in the PR does this positioning of ir
irblaze
So the walls aren't sensed. Now its upto you to decide the model of mouse you want to use. But if you ask me, I would say putting ir in the middle of the mouse and detecting already traversed wall when moving half forward makes the half forward function useless in terms of productivity. 
This is the actual use of half forward
halfsense
The mouse doesn't go completely into the box, but just goes halfway and decides that going there is useless and so it saves time. With the current model the half forward, the mouse has to go entirely into the cell, so this leaves no reason for using half forward. And another thing, you have to make the half forward an command that can accept distance, because suppose i am in this position
path
With the current simulator i would have to go like this,

```
HalfForward = 1
Forward = 5
HalfForward = 1
```
But if i used distance on half forward i could do, 

`HalfForward =12`

These are what i noticed, i already fixed this in the PR, and if you need help you could check the code their and incorporate them yourself.

**Comment by mackorone** [comment]
@Blaze-Leo this feedback is fantastic, thank you for writing it up! I'm busy until the end of this week but I'll take a look at your PR and try to incorporate your suggestions this weekend.

**Comment by mackorone** [comment]
@Blaze-Leo, sorry for the confusion. The `moveForward` and `moveForwardHalf` commands work in both the cardinal and ordinal (diagonal) directions. So to move diagonally, you could call `turnRight45` and `moveForward`, for example. Does this API work for you?

**Comment by mackorone** [comment]
> a diagonal movement is a different distance from forward

Indeed!

> there has to be a different function for diagonal otherwise the user will not be able to use the simulator

I totally get that we don't want users to wrongly assume that moving forward in a cardinal direction is the same as moving forward in an ordinal direction, but I decided to reuse the `moveForward` method mainly to avoid confusing or surprising behavior in the case when the wrong "move forward" method is invoked. It's not that there exists no reasonable behavior for that case; it's that, regardless of what behavior we chose, it would probably be unintuitive. I decided that repurposing the `moveForward` method was the lessor of two evils.

Regarding your concern about not being able to use the simulator: can you define two helper functions. e.g. `moveForwardCardinal` and `moveForwardOrdinal`, and call them and not `moveForward` from within your algorithm code? Presumably you could implement the behavior that you want in those helper functions.

> I still don't understand why you choose to do the changes yourself

It's a fair question. It was simply a matter of what was easiest and fastest for me. I needed to make sure that I understood your changes fully, and that they were incorporated into the codebase sustainably. We could've gone through a few more rounds of code review, but it was just easier to do it myself. Of course, the credit for these changes is all yours! I certainly couldn't have made them without your guidance!

**Comment by Blaze-Leo** [comment]
I would like to make a declaration before starting my comment. Our goal is to help the user be successful in his journey of building a micromouse, making the code free of errors by using patch up work is not our goal. If the user by mistake calls forward instead of diagonal, our task is not to make the code somehow work. Its our duty to point out all the flaws in the user's code so that after he has made a successful software and implements it with his hardware, and by any chance receives any errors, he doesn't have to look back into the software for bugs. So as a way to make the user's software completely bug-free we must not make use of moveForward for diagonal as well as straight movement.

Using the names `moveForwardOrdinal` and `moveForwardCardinal` is a bit lengthy as well as technical, I would still suggest the naming convention of `moveForward` and `moveDiagonal`. And these two shall not be helper functions but the functions itself to be called.

> Regarding your concern about not being able to use the simulator: can you define two helper functions. e.g. moveForwardCardinal and moveForwardOrdinal, and call them and not moveForward from within your algorithm code? Presumably you could implement the behavior that you want in those helper functions.

I kind of feel like I will have to explain my own code now. In fact I have been thinking, it doesn't have to be you, but whoever would have asked me any questions about my code, all i could offer is the explanations. Its because I had the best ideas put in the code for which i built it. So if anyone ever asks me for new suggestions, my ideas will be limited to my code, because that's the best I could think.

In the codebase the two functions `moveForward` and `moveDiagonal` will be referred to as different functions, but inside the function the only different part will be the distance to be traveled. With diagonal movement comes problem of wall sensing. I have made some calculations and with that my diagonal wall sensing is perfect. You could look in the wallDiagonal or something fuction, I don't remember accurately, (just below isWallFront function) and I can guarantee you will have doubts about what the heck is going on that function, and if you do, you may ask me.

diag

As you can see, in position 1 we have to check that the mouse doesn't go to the the top, in 2 we have to check that the mouse doesn't go to right and down, and at 3 it doesn't go to top again. All I wanted to explain that since the diagonal movements are different from forward in respect to wall sensing, we must create another wall sensing function dealing with the crash sensing. So putting helper functions is not the ideal solution, instead creating new functions should be the goal. 

I would again say, your mms is not made to be a user-friendly software, its supposed to be a hard game, (something like Hollow Knight, took me 3 months to finish). The user's code must go through rigorous testing at must work in all conditions. If the user doesn't want to call a `moveDiagonal` method and somehow makes his code work by `moveForward`, we must not allow that. 

As an example I will state my hardships, and I would say every simulator user must go through these hardships. I created three versions of the code, every version uses a different algorithm. And under each version I have at least 10 types of code, out which every one has the same alogrithm but uses different way of approach. And every stable software I have created, has been tested with the 480 maps provided in your Readme file, and a at least after 6 months of coding I have come up with my final software which can solve every map in the best way I could imagine. Had your simulator been user-friendly, I could have been finished in just 1 version, but that wouldn't have been ideal in a competition. The goal of mms is such, that in any competition, people who have used mms must be at a significant advantage from people who haven't.

And another thing, seems like your simulator cannot identify mazes less that 16x16 in size. When the input is a 5x10 maze, the simulator fills the rest of 16x16 squares as not reachable (infinity when displaying flood fill). You might want to fix this.

**Comment by mackorone** [comment]
> we must not make use of moveForward for diagonal as well as straight movement

I think we'll just have to agree to disagree for now.

Separately, I have a few quick updates:
1. There are now five new wall-checking methods
   1. `wallBack`
   2. `wallFrontLeft`
   3. `wallFrontRight`
   4. `wallBackLeft`
   5. `wallBackRight`
1. The wall-checking methods now take an optional `halfStepsAway` argument. By default `halfStepsAway=1`, which means that the sensor length will be exactly one half-step. By specifying `halfStepsAway=2` and above, the mouse can check for the existence of walls further away than the current cell. It's a little bit confusing because "half step" isn't a fix distance - it depends on what direction the mouse if facing (similar to the semantics of `moveForward`). However, it should be enough functionality to get you unblocked.
1. I updated the `mms-python` repo to add support for the new methods and arguments.

When you get a chance, I'd love to hear if these changes solve your problem (regarding reading wall values without completely entering a cell). Additionally, can you please send a screenshot of this problem? I'd love to take a look.

>  When the input is a 5x10 maze, the simulator fills the rest of 16x16 squares as not reachable (infinity when displaying flood fill). You might want to fix this.

**Comment by Blaze-Leo** [comment]
I didn't test the functions cause I think they where really simple considering you have a wall function which adds the half steps value and then returns the walls. That does solve my problem as well as adds new functionality. As for me, the default half steps =1 will make my code backwards compatible, although I might have to change the API.

And about the 5x10 map, there were trailing blank-spaces in the map which caused the simulator either shutdown or show a 16 x16 blank map( the default map). But with the spaces removed it now works fine.

And about diagonal, I tried turnRight45 then moveForward 1, but that creates cannot moveForward error. Can you give an example for the diagonal movement?

**Comment by mackorone** [comment]
> I tried turnRight45 then moveForward 1

I think that a diagonal movement requires that you be on the edge of a cell. So try this, assuming that there are no walls obstructing:
1. `moveForwardHalf`
1. `turnRight45`
1. `moveForward`

> there were trailing blank-spaces in the map

Ah, ok, thanks for clarifying.

> could you please continue to update the mms-java

Unfortunately, I won't have time to do this in the near term. That said, if you're willing, I'd be happy to review a PR!

**Comment by Blaze-Leo** [comment]
Ah, actually I thought that if there are no walls I could just go anywhere. Yes, with half forward move at first the diagonal works fine.

> 
if you're willing, I'd be happy to review a PR!

I didn't understand what you meant.

**Pull request closed without merging**